File: ports/php/src/LombokAlgoritma/ML/VectorOps.php
<?php

// LombokAlgoritma — PHP Vector operations
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

declare(strict_types=1);

namespace LombokAlgoritma\ML;

/** Dense vector math with left-to-right (sequential) summation, matching the other ports. */
final class VectorOps
{
    /**
     * @param list<int|float> $a
     * @param list<int|float> $b
     * @throws \InvalidArgumentException on length mismatch
     */
    public static function dot(array $a, array $b): float
    {
        self::sameLength($a, $b);
        $s = 0.0;
        foreach ($a as $i => $x) {
            $s += $x * $b[$i];
        }
        return $s;
    }

    /** @param list<int|float> $v */
    public static function norm(array $v): float
    {
        $s = 0.0;
        foreach ($v as $x) {
            $s += $x * $x;
        }
        return sqrt($s);
    }

    /**
     * Cosine similarity; 0 when either vector is zero.
     *
     * @param list<int|float> $a
     * @param list<int|float> $b
     */
    public static function cosine(array $a, array $b): float
    {
        $na = self::norm($a);
        $nb = self::norm($b);
        return ($na == 0.0 || $nb == 0.0) ? 0.0 : self::dot($a, $b) / ($na * $nb);
    }

    /**
     * Euclidean distance.
     *
     * @param list<int|float> $a
     * @param list<int|float> $b
     */
    public static function l2(array $a, array $b): float
    {
        self::sameLength($a, $b);
        $s = 0.0;
        foreach ($a as $i => $x) {
            $d = $x - $b[$i];
            $s += $d * $d;
        }
        return sqrt($s);
    }

    /**
     * @param list<int|float> $v
     * @return list<float>
     */
    public static function normalize(array $v): array
    {
        $n = self::norm($v);
        if ($n == 0.0) {
            return array_fill(0, count($v), 0.0);
        }
        return array_map(static fn (int|float $x): float => $x / $n, $v);
    }

    /**
     * @param list<int|float> $a
     * @param list<int|float> $b
     */
    private static function sameLength(array $a, array $b): void
    {
        if (count($a) !== count($b)) {
            throw new \InvalidArgumentException('vectors must have equal length');
        }
    }
}

File: csharp/src/ML/VectorOps.cs
// LombokAlgoritma — C# ML/Vector Operations
// DIPAKAI oleh LombokVector (C# port)
// Apache-2.0 — @codinglombok
using System;

namespace CodingLombok.LombokAlgoritma.ML;

/// <summary>Vector similarity and distance functions (SPEC §13.1).</summary>
public static class VectorOps
{
    /// <summary>Dot product Σ aᵢ·bᵢ (summed left to right).</summary>
    public static double DotProduct(double[] a, double[] b)
    {
        var sum = 0.0;
        for (var i = 0; i < a.Length; i++) sum += a[i] * b[i];
        return sum;
    }

    /// <summary>Euclidean norm √Σ vᵢ².</summary>
    public static double L2Norm(double[] v)
    {
        var sum = 0.0;
        foreach (var x in v) sum += x * x;
        return Math.Sqrt(sum);
    }

    /// <summary>Cosine similarity dot(a,b)/(‖a‖·‖b‖); 0 when either vector is zero.</summary>
    public static double CosineSimilarity(double[] a, double[] b)
    {
        var na = L2Norm(a); var nb = L2Norm(b);
        return na == 0 || nb == 0 ? 0.0 : DotProduct(a, b) / (na * nb);
    }

    /// <summary>Euclidean distance √Σ(aᵢ−bᵢ)².</summary>
    public static double L2Distance(double[] a, double[] b)
    {
        var sum = 0.0;
        for (var i = 0; i < a.Length; i++) { var d = a[i] - b[i]; sum += d * d; }
        return Math.Sqrt(sum);
    }

    /// <summary>Unit vector v/‖v‖ (zero vector stays zero).</summary>
    public static double[] Normalize(double[] v)
    {
        var n = L2Norm(v);
        if (n == 0) return new double[v.Length];
        var out_ = new double[v.Length];
        for (var i = 0; i < v.Length; i++) out_[i] = v[i] / n;
        return out_;
    }
}
